compiler/resolution/cullOverReferences.cpp

```diff
@@ -185,54 +184,101 @@ void cullOverReferences() {
+
+    // Make sure that the context call only has 2 options.
+    INT_ASSERT(cc->options.length == 2);
+
+    FnSymbol* fn = refCall->isResolved();
+    INT_ASSERT(fn);
+    FnSymbol* valueFn = valueCall->isResolved();
+    INT_ASSERT(valueFn);
+
+    bool useValueCall;
+
+    CallExpr* move = NULL; // set if the call is in a PRIM_MOVE
+    SymExpr* lhs = NULL; // lhs if call is in a PRIM_MOVE
+
+    // Decide whether to use the value call or the ref call.
+    // Always leave the ref call for iterators.
+    // (It would be an improvement to choose the appropriate one
+    //  based upon how the iterator is used, but such a feature
+    //  would require specific support for iterators since yielding
+    //  is not the same as returning.)
+    move = toCallExpr(cc->parentExpr);
+    if (fn->isIterator())
+      useValueCall = false;
+    else if (move) {
+      INT_ASSERT(move->isPrimitive(PRIM_MOVE));
+
+      lhs = toSymExpr(move->get(1));
+
+      INT_ASSERT(lhs);
+
+      // Should we switch to the by-value form?
+      useValueCall = shouldUseByValueFunction(fn, lhs, defMap, useMap);
+    } else {
+      // e.g. array access in own statement like this:
+      //   A(i)
+      // should use 'getter'
+      // MPF - note 2016-01: this code does not seem to be triggered
+      // in the present compiler.
+      useValueCall = true;
+    }
+
+    valueCall->remove();
+    refCall->remove();
+
+    if (useValueCall) {
+      // Replace the ContextCallExpr with the value call
+      cc->replace(valueCall);
+
+      // Adjust the AST around the value call to include
+      // a temporary to receive the value.
+
+      // Adjust code to use value return version.
+      // The other option is that retTag is RET_CONST_REF,
+      // in which case no further adjustment is necessary.
+      if (move && valueFn->retTag == RET_VALUE) {
+        SET_LINENO(move);
+        // Generate a value temp to receive the value
+        VarSymbol* tmp  = newTemp(valueFn->retType);
+        move->insertBefore(new DefExpr(tmp));
+
+        if (requiresImplicitDestroy(valueCall)) {
+          if (isString(valueFn->retType) == false) {
+            tmp->addFlag(FLAG_INSERT_AUTO_COPY);
+            tmp->addFlag(FLAG_INSERT_AUTO_DESTROY);
+          } else {
+            tmp->addFlag(FLAG_INSERT_AUTO_DESTROY);
           }
+        }
+
+        if (lhs && useMap.get(lhs->var) && useMap.get(lhs->var)->n > 0) {
+          // If the LHS was used, set it based on the
+          // new temporary (which is the function return value)
+          move->insertAfter(new CallExpr(PRIM_MOVE,
+                                         lhs->var,
+                                         new CallExpr(PRIM_ADDR_OF, tmp)));
         } else {
-          INT_FATAL(call, "unexpected case");
+          // If the LHS was not used,
+          // remove the old definition point since we have
+          // provided a new one above.
+          lhs->var->defPoint->remove();
```

Comment:
I don't think that would be a good idea. If you call a function in a ref-pair, like this

refPairFunction();

and you don't use the result at all, this code could be triggered. The way I have it now,
it will still use the value version in that case. That makes sense to me, since you wouldn't
want to (e.g.) add Sparse array elements with such a call.